File: lib/codeintel/autoindex/inference/rust.go
package inference

import (
	"path/filepath"

	"github.com/grafana/regexp"

	"github.com/sourcegraph/sourcegraph/lib/codeintel/autoindex/config"
)

func InferRustIndexJobs(gitserver GitClient, paths []string) (indexes []config.IndexJob) {
	for _, path := range paths {
		if canIndexRustPath(path) {
			// We only create one IndexJob, as the complexity of indexing
			// large projects, such as those with workspaces, is handled
			// in lsif-rust.
			indexes = append(indexes, config.IndexJob{
				Indexer:     "sourcegraph/lsif-rust",
				IndexerArgs: []string{"lsif-rust", "index"},
				Outfile:     "dump.lsif",
				Root:        "",
				Steps:       []config.DockerStep{},
			})
			break
		}
	}
	return indexes
}

func RustPatterns() []*regexp.Regexp {
	return []*regexp.Regexp{
		suffixPattern(rawPattern("Cargo.toml")),
	}
}

func canIndexRustPath(path string) bool {
	return filepath.Base(path) == "Cargo.toml"
}

File: internal/codeintel/autoindexing/internal/inference/lua/rust.lua
local pattern = require "sg.autoindex.patterns"
local recognizers = require "sg.recognizers"

local indexer = "sourcegraph/lsif-rust"
local outfile = "dump.lsif"

return recognizers.path_recognizer {
  patterns = {
    pattern.new_path_basename "Cargo.toml",
  },

  -- Invoked when Cargo.toml exists anywhere in repository
  generate = function(_, paths)
    return {
      steps = {},
      root = "",
      indexer = indexer,
      indexer_args = { "lsif-rust", "index" },
      outfile = outfile,
    }
  end,
}
